File: Javascript/src/01.AlgorithmicToolbox/Week3GreedyAlgorithms/maxDotProduct.js
/**
 * Problem 3 - Maximum Advertisement Revenue
 * Max Dot Product
 * @param {Array.<number>} sequence1 first sequence of numbers
 * @param {Array.<number>} sequence2 second sequence of numbers
 * @returns maximum dot product of the two sequences
 */
export function maxDotProduct(sequence1, sequence2) {
  sequence1.sort();
  sequence2.sort();
  return sequence1
    .map((value, index) => value * sequence2[index])
    .reduce((acc, cur) => acc + cur);
}

File: Rust/src/_01_algorithmic_toolbox/week3_greedy_algorithms/max_dot_product.rs
/// Problem 3 - Maximum Advertisement Revenue
///
/// Returns the maximum dot product of two given sequence of integers of equal lengths
pub fn max_dot_product(mut sequence1: Vec<i64>, mut sequence2: Vec<i64>) -> i64 {
    sequence1.sort();
    sequence2.sort();
    sequence1
        .iter()
        .zip(sequence2.iter())
        .map(|(first, second)| first * second)
        .sum()
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn returns_max_dot_product_of_two_sequences() {
        assert_eq!(max_dot_product(vec![23], vec![39]), 897);
        assert_eq!(max_dot_product(vec![1, 3, -5], vec![-2, 4, 1]), 23);
    }
}
